### DISCUSSION TITLE
Multiple rendering - Render template once per CSV record

### DISCUSSION THREAD

**Discussion by safferyj** [hairyhenderson/gomplate#2124] [Q&A] [answered]
My use case is to make multiple API calls using a JSON template document with substituted values from each line of a CSV data file. I am on Windows and not a Go programmer but hoping that gomplate will be a suitable tool for this.

For example, if I have a file test.json as:

```
{
    "a": "1",
    "b": "2",
    "c": "hello"
}
```

And can make an API call with:

`type test.json | curl --request POST --header "Content-Type: application/json" --data-binary @- http://api/`

Then, can I make a single call to gomplate with a data file like this

```
a,b,c
1,2,hello
3,4,world
5,6,yippee
7,8,yay
```

and a template file something like this

```
{
    "a": "{{.a}}",
    "b": "{{.b}}",
    "c": "{{.c}}"
}
```

and achieve 4 calls to the API?

I have spent some time trying various options but have not had any success. Many thanks.

**Comment by hairyhenderson** [comment]
Hi @safferyj, in general gomplate doesn't support `POST`ing to HTTP datasources. It uses `GET` for those.

But, if the API was constructed differently and could accept the payload as URL query parameters you could accomplish it by feeding a subpath argument to the `datasource`/`ds` function - see https://docs.gomplate.ca/functions/data/#datasource

**Reply by safferyj** [reply]
Thanks. I was thinking of using the `--exec-pipe` feature to pipe the rendered template to `curl`. If doing this, is it possible somehow to render the template multiple times with the data for each occurrence coming from a line read from CSV?

**Reply by hairyhenderson** [reply]
The process called by `--exec-pipe` is only called once. To accomplish what you're looking for I think you'd need to just wrap the command in a shell script and loop it.

**Reply by safferyj** [reply]
Thanks. I arrived at the gomplate and curl plugin solution posted below.

**Comment by safferyj** [comment, marked as answer]
On Windows, with

```
type template.tmpl
{
    "a": "{{.a}}",
    "b": "{{.b}}",
    "c": "{{.c}}"
}
```
and 

```
type data.csv
a,b,c
1,2,hello
3,4,world
5,6,yippee
7,8,yay
```
then

```
gomplate --plugin curl=curl.exe -t template.tmpl -d data.csv -i "{{$data := include `data` | csvByRow}}{{range $data}}{{$render := tmpl.Exec `template.tmpl` .}}{{curl `--silent` `--json` $render `http://api`}}{{end}}"
```
will make the following four distinct POST API calls

```
{
    "a": "1",
    "b": "2",
    "c": "hello"
}
```

```
{
    "a": "3",
    "b": "4",
    "c": "world"
}
```

```
{
    "a": "5",
    "b": "6",
    "c": "yippee"
}
```

```
{
    "a": "7",
    "b": "8",
    "c": "yay"
}
```
